### PR TITLE
DeepSSM 

### PR DESCRIPTION
**Pull request by jadie1**
This pull request:
- adds DataAugmentationUtils to Python packages (issue #581 and issue #625)
- adds DeepSSMUtils to Python packages (issue #581)
- updates conda_installs.sh to include DataAugmentationUtils, DeepSSMUtils, Pytorch (GPU if available otherwise CPU version), and bokeh (issue #481)
- adds a new use case: deep_ssm.py (issue #551)
- adds a new mesh function: SurfaceToSurfaceDistance (issue #527)
- adds documentaton (issue #552)
	- docs/deep-learning/deep-ssm.md
	- docs/deep-learning/data-augmentaion.md
	- docs/use-cases/deep-ssm-femur.md

### CODE REVIEW

**Comment by jadie1** [comment]
To test I recommend:
1. Run conda_installs.sh
2. Run build_dependencies.sh and cmake again if you haven't already for the VTK 9.0 upgrade
3. Run the use case tiny test:
```
cd ShapeWorks/Examples/Python/
python RunUseCase.py --use_case deep_ssm --tiny_test
```
If you get the Pytorch running on CPU error, then conda_installs.sh picked the wrong CUDA version. Install the correct one with:
`conda install -n shapeworks pytorch torchvision cudatoolkit=CORRECT_VERSION_NUM -c pytorch`
4. Check over the documentation (@riddhishb I could use your advice here)

You can also run the full use case, it takes about a day to run. The tiny_test does not skip any steps but the results are bad so to see the accuracy of the results the full use case needs to be run. 

I tested the surface to surface distance tool pretty thoroughly, here is how to use it:
`SurfaceToSurfaceDistance --inputA path_to_input_mesh_A --inputB path_to_input_mesh_B --outputA path_to_output_mesh_A --outputB path_to_output_mesh_B`

The input meshes need to be VTK. The output meshes are the same as the input but they have a scalar field on them with the distance value, this can be visualized as a heatmap in ParaView. This is just the base functionality, SurfaceToSurfaceDistance will be updated next release to be more flexible and do more (issue #708)

File: Python/DataAugmentationUtilsPackage/DataAugmentationUtils/Sampler.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+		return mixture_num
+	def sample(self):
+		sample = self.GMM.sample(1)[0][0]
+		closest_index = getClosest(sample, self.embedded_matrix)
+		return sample, closest_index
+
+# instance of Sampler class that uses kernel density estimate
+class KDE_Sampler(Sampler):
+	def fit(self, embedded_matrix):
+		print("Fittting KDE...")
```
Comment (akenmorris) [inline Python/DataAugmentationUtilsPackage/DataAugmentationUtils/Sampler.py:65]: Typo, Fittting -> Fitting.

File: docs/deep-learning/deep-ssm.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
+## ShapeWorks DeepSSM
+DeepSSM is a convolutional neural network (CNN) model that can provide particles directly from unsegmented images after it has been trained. This documentation provides an overview of the DeepSSM process, for a full explanation see: [DeepSSM: A Deep Learning Framework for Statistical
```
Comment (riddhishb) [inline docs/deep-learning/deep-ssm.md:2]: DeepSSM is the framework which includes data-augmentation and then the prediciton is made by a CNN model trained on image,correspondence pairs.

File: docs/deep-learning/deep-ssm.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+Shape Modeling from Raw Images](https://arxiv.org/abs/1810.00111).
+
+The input to the DeepSSM network are unsegmented 3D images of the anatomy of interest and the output are the particle-based shape models. DeepSSM requires training examples of image/PBM pairs which are generated via the traditional Shapeworks grooming and optimization pipeline or otherwise. Once the network has been trained on these examples it can predict the PBM of unseen examples given only images, bypassing the need for labor intensive segmentation, grooming, and optimization parameter tuning.
```
Comment (riddhishb) [inline docs/deep-learning/deep-ssm.md:5]: otherwise => other Particle distribution models

File: docs/deep-learning/deep-ssm.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+The DeepSSM network is implemented in PyTorch and requires a GPU to run efficiently. 
+
+### DeepSSM Steps 
+
+#### 1. Data Augmentation
+
+The first step to creating a DeepSSM model is generating training data. Deep networks require thousands of training instances and so because medical imaging data is typically limited, data augmentation is necessary. The data augmentation process is described here:  [Running Data Augmentation](DataAugmentation.md).
+
+The data augmentation process involves reducing the PBM's to a low dimensional space via Principal Component Analysis (PCA), preserving a chosen percent of the variation. The PCA scores are saved and used as the labels for DeepSSM prediction. The PCA scores are deterministically mapped back to the PDM using the eigenvalues and vectors once the DeepSSM model makes a prediction.
```
Comment (riddhishb) [inline docs/deep-learning/deep-ssm.md:20]: PBM ? or PDM ?

File: docs/deep-learning/deep-ssm.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+![DeepSSM Architecture](../img/deep-learning/Architecture.png)
+
+#### 4. Testing
+
+The trained model is then used to predict the PCA score from the images in the test set. These PCA scores are then un-whitened and mapped back to the particle coordinates using the eigenvalues and eigenvectors from PCA. Thus a PDM is acquired for each test image.
+
+#### 5. Evaluation
+
+To evaluate the accuracy of DeepSSM output, we compare a mesh created from segmentation to a mesh created from the predicted PDM. To obtain the original mesh, we use the ShapeWorks MeshFromDistanceTransforms command to get a mesh from the distance transform created from the true segmentation. To obtain the predicted mesh, we use the ShapeWorks ReconstructSurface command with the mean and predicted particles to reconstruct a surface.
```
Comment (riddhishb) [inline docs/deep-learning/deep-ssm.md:38]: command format

File: docs/deep-learning/deep-ssm.md (hunk 3 above)
Comment (riddhishb) [inline docs/deep-learning/deep-ssm.md:5]: PBM/PDM ? 
Also the first time the abbreviation is declared give the full form, after that you can use it freely/

File: docs/use-cases/deep-ssm-femur.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
+# DeepSSM: ShapeWorks Deep Learning on Femur Data
```
Comment (riddhishb) [inline docs/use-cases/deep-ssm-femur.md:1]: I think this use case should be in use-cases/deep-learning, right ?

**Review by riddhishb — CHANGES_REQUESTED** [review]
Hi @jadie1 I have reviewed the documentation, the comments are in the diff.

**Comment by akenmorris** [comment]
This is the output I get on the tinytest on mac, is this right?

Screen Shot 2020-09-28 at 2 31 58 PM

**Comment by jadie1** [comment]
@akenmorris Yeah that's right! This is just a preliminary visualization, we have an issue to improve it for the next release.

File: docs/use-cases/deep-ssm-femur.md (hunk 6 above)
Comment (jadie1) [reply docs/use-cases/deep-ssm-femur.md:1]: @sheryjoe What do you think? Should we have a deep-learning folder inside the use-cases folder?

**Comment by jadie1** [comment]
@sheryjoe I have addressed most of Riddhishs comments on the documentation now if you'd like to look over it.

**Comment by riddhishb** [comment]
I get this, different from Alan's plot
Screenshot_2020-09-28_19-54-58

**Comment by jadie1** [comment]
Looks good. The real examples (blue) are the same and the augmented (red) are sampled randomly so we expect them to be different.

**Comment by iyerkrithika21** [comment]
image
Output on windows.

**Comment by akenmorris** [comment]
Ok, the windows build is fixed.

**Comment by akenmorris** [comment]
Are @riddhishb ’s requested changes resolved?

**Comment by jadie1** [comment]
@akenmorris working on them now, then it will be ready to merge.

File: docs/deep-learning/data-augmentation.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+#### Multivariate Gaussian Distribution
+
+To fit multivariate Gaussian distribution (which can be seen in Step 2 above), the probability density function is parameterized by the mean of and covariance of the embedded data matrix. This normal distribution is then randomly sampled from to get new examples. The closest real example to each sampled point is found by calculating the Mahalanobis distance. 
+
+#### Mixture of Multivariate Gaussians Distribution
+
+A Gaussian mixture model can provide a more appropriate probability density function when the embedded data distribution is mutli-modal. To fit a mixture model, first we cluster on the embedded data and select the optimal number of clusters by minimizing Akaike information criterion (AIC) and Bayesian information criterion (BIC). This number determines how many Gaussian distributions (or components) should be used. Next the expectation-maximization (EM) algorithm is used to fit a mixture-of-Gaussian model with this number of components. This distribution can then be randomly sampled from and the closest real example is chosen using Mahalanobis distance.
```
[the remark this answers is not in the document]
Comment (jadie1) [reply docs/deep-learning/data-augmentation.md:37]: It is both, we find the number of components that minimize AIC and the number of components that minimize BIC and average them.

**Comment by jadie1** [comment]
@sheryjoe @akenmorris I finished making the updates Riddhish suggested for the documentation and I tested everything including the surface to surface distance tool with VTK 8.2 and it is all working (thank you Alan!) I think this is ready to merge!

**Comment by akenmorris** [comment]
Great. I will just wait for the ok from the checks and then merge.
